src/ipld/cid_vec.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::utils::cid::{CidVariant, BLAKE2B256_SIZE};
+use cid::{
+    multihash::{self, Code::Blake2b256},
+    Cid,
+};
+use fvm_ipld_encoding::DAG_CBOR;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+// CidVec takes advantage of the fact that the V1 DAG-CBOR Blake2b-256 variant (which can be stored in 32 bytes vs 96 bytes for a `Cid` type) is +99.99% of all CIDs. CidVec defaults to the `Vec<[u8; BLAKE2B256_SIZE]>` type, only using the more expensive `Vec<Cid>` type when necessary.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum CidVec {
+    V1Cids(Vec<[u8; BLAKE2B256_SIZE]>),
+    AllCids(Vec<Cid>),
+}
+
+impl Serialize for CidVec {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.cids().serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for CidVec {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(Self::from(<Vec<Cid>>::deserialize(deserializer)?))
+    }
+}
+
+impl Default for CidVec {
+    fn default() -> Self {
+        Self::V1Cids(Vec::new())
+    }
+}
+
+impl FromIterator<Cid> for CidVec {
+    fn from_iter<T: IntoIterator<Item = Cid>>(iter: T) -> Self {
+        let mut vec = Self::new();
+        for i in iter {
+            vec.push(i);
+        }
+        vec
+    }
+}
+
+impl From<Vec<Cid>> for CidVec {
+    fn from(vec: Vec<Cid>) -> Self {
+        // Converts `Vec<Cid>` to `CidVec::V1Cids` if possible; otherwise, converts to `CidVec::AllCids`.
+        let mut cid_vec = CidVec::new();
+        for cid in vec {
+            cid_vec.push(cid);
+        }
+        cid_vec
+    }
+}
+
+impl From<&[Cid]> for CidVec {
+    fn from(vec: &[Cid]) -> Self {
+        vec.iter().cloned().collect()
+    }
+}
```

Comment:
I don't think this should exist - users should control the clones, and there's a FromIterator implementation already

Reply:
Removed.